### DISCUSSION TITLE
Any changes to LV426PT2 NPCs movment speed breaks NPC movement and objective prompt vanishes

### DISCUSSION THREAD

**Discussion by ghramsey** [MattFiler/LegendPlugin#148] [General]
I'm a bit frustrated with this.  I managed to happily speed up the NPCs in LV426.  They run instead of just turtling along.
This is fine until the transition to the 2nd half.  Soon as the transition finishes they all just freeze in place.  The only thing 
I've set is move=3 which is run in all of their cmd_goto settings.   This works in one  but not the other. That's very confusing.

**Comment by MattFiler** [comment]
Odd! I wonder if something else is overriding it somewhere? Are there any other locomotion modifier entities pointing to them?

**Reply by ghramsey** [reply]
IDK what links where between the two maps, but anytime I speed up Pt1, P2 breaks right from the start.

I'm going to have to painstakingly trial-and-error it with small edits.  Right now I only sped up the banter at the start
and the boost ups from Heyst and I set the cage animation speed for the winch to 20 and still got frozen characters.

Going to be a long time working on this at this rate.
 I see why other mods just skip the mission, but I enjoy the challenge of this.

**Reply by ghramsey** [reply]
I seem to have found something, but I don't l know yet why.

The break is becasuse the mission goal 'find the source of the signal' is not showing up.
As soon as the prompt shows the NPCs all begin walking forward.  

When I use an unedited Pt2 commands.pak the prompt shows up.  
If I edit Pt1 or Pt2 to make the characters move faster then the prompt breaks.

**Reply by ghramsey** [reply]
Question and I expect this is a red herring:
 Is there any debug functionality in the game that commandseditor could tap to find and debug hang ups like this?
 Or any way to make a log file output ?
 
 I am slowly tracking this to my adding 'PlaySpeed' settings to CMD animations  and 'playspeed' on the CAGE animations.   
 But they're not or I don't think they're related to the initial load into the level other than probably being referenced.
 I cannot even figure where at this point.  Everytime I edit the playspeeds it hangs up the movement and the objective fails to appear.
 
 I think the toughest part of debugging this is the fact that it hangs if I try to reload the save after the transition. 
 I thought it was a checkpoint but it won't load there.  Not sure why. There is an auto checkpoint so I thought it 
 would be designed to be able to start there but guess not.   Everytime I want to test I have to go to beginning of the
 level.

**Reply by MattFiler** [reply]
Yeah I recognise that debugging is a chore currently. I'm working on reimplementing the `DebugText` entity outputs which would allow you to output data to a console which should help, among other things. Nothing yet though - sorry! Trial and error is the key for now :D

**Reply by ghramsey** [reply]
I cannot find what link in the chain is breaking.  Seems to be related to the move functions minimum and maximum in the cmd goto commands for the 3 NPCS.  They have similar functions in Pt1 and those work flawlessly.  
I'm going to see what else I can edit instead of the movement.  I've been saving incrementally of the settings that do work.
So far all of the running in part 1 works fine. I can shorten the dialog by just removing reference to it.   Then the transition 
just breaks unless I just modify the chr movement speed that lets me run.

**Reply by ghramsey** [reply]
I think I just saw something a moment.  While loading in the game seemed like it might have 
loaded the prompt too fast to be seen.  And again the characters break.  So it's breaking because a delay is missing.  
Hmmmm.

But what's odd is I reduced the change to only the cmd animation for the mantle Lewis gives us and interaction with
the beacon.   I'm still perplexed why these are breaking the mission load-in.

**Reply by ghramsey** [reply]
So the objective that is failing to fire is "find the source of the signal".  
=
[A1_M1302_OBJ_1803]
{Find the source of the signal}
=
I've searched through all of the pre mission and mission start and I cannot locate where it's being referenced.
I can find where it's defined as a PRIOBJ but I cannot locate where it's being called from to back track it.
I'm afraid I'm still  a rank noob at understanding how these scripts flow together. 

Any help with locating where this is being called or perhaps not being called? 
 It's at least a symptom of the issue.

**Reply by MattFiler** [reply]
I'll take a look this morning

**Reply by ghramsey** [reply]
I appreciated whatever you can do.     I am still not sure if the issue is in part2 or carrying over from Pt1 to Pt2.

I just discovered something new though.  I've been under the assumption that the NPCs were stuck since they were not following me all the way to the Space Jockey.  Seems that was false assumption on my part. They were either merely delayed to "normal" speed (once past the first triggerbox inside the derilict there's a collision barrier so I could not look)  or teleported on some timer.    I went away at the ledge where I need Meeks to boost me. He was nowhere in sight the whole way to that point.  I came back probably 10m later and he was there crouched and waiting.  So maybe like in M2 where Axel teleports to you when you get past a certain distance.  All of the characters have a goto set with a setting of 6 as well as min and max move speed settings.  So when I wake up later I'm going to set all those movement from 1 which is walk to 3 which is run.  And see if I catch them running up behind me after time.    It's bedtime now though; 145am.

This finding though doesn't negate the initial issue of them being stuck upon transition to the 2nd map.
Maybe I'll live with it if they do eventually teleport though.

**Reply by MattFiler** [reply]
The objective is triggered by the `SetPrimaryObjective` entity in the `SCRIPT_STORYMISSION` composite called `PRIMOBJ_FindSourceOfSignal`. It's not hooked up to anything but I assume it's just being triggered on reset.

RE how the NPCs are moving, they're just following a spline by the looks of it & going up to 4 points ahead on the spline from the player's position. Have you tried using `CHR_LocomotionModifier` or `NPC_SetLocomotionTargetSpeed` to set their walk speed? The former has `Must_Run` which is a bool, and also has individual options for being able to walk etc which you could disable.

**Reply by ghramsey** [reply]
I confirmed that just waiting for Meeks does seems to work around the whole issue.  He shows up over a few minutes time even if 
I set the characters to run through the CMD_FollowUsingJobs fastest_allowed_move_type and slowest_allowed_move-type 
ENUMs set MOVE to 3 which is run.  This is what I set on the first half and they all sprint along.  They don't seem to be obeying this
setting in the 2nd map.  Seems to point to your original comment about something overriding it.

So that objective is being satisfied upon entering the 2nd part of the mission; the 2nd map.  They move forward on a spline and it's linked to a movement Jobs.  Any clue as to what sets the movement Job into motion?  That is what is breaking. The objective is a trigger for their movement.  If it's on_reset there's something else disconnecting.

I used CHR_Locomotionmodifer can_run set to checked on Marlow which lets the player run.
I'm not aware of that NPC_SetLocomotionTargetSpeed setting. Where do I look for it?

**Reply by ghramsey** [reply]
> The objective is triggered by the `SetPrimaryObjective` entity in the `SCRIPT_STORYMISSION` composite called `PRIMOBJ_FindSourceOfSignal`. It's not hooked up to anything but I assume it's just being triggered on reset.
I think the PRIMOBJ in PT2 is a success carried over from the previous map.  Since the PRIMOBJ is Find the Signal from the start of the whole level.

In any case this, entire mission is being hosed whenever I edit PT2.  Even a tiny change now breaks the NPCs' initial movement.
I reverted to a stock default, unaltered LV426_PT2 file and edited only two things: The CHR Locomotion modifier to allow running 
on the player which usually works without issue and the speechscript for Marlow when he gets close to the beacon.  The same issue
 is still happening for this minor change as well.

And I just experienced an issue where it was crashing the game just loading into PT2 map with the vanilla game files.
I reverted the Commands.Pak and Models.MVR which were edited but another file is edited too; REDS.BIN.
What does that file do ?  Maybe it's related.

**Comment by ghramsey** [comment]
The CHR_LocomotionModifier is setup for the player under   archetype/npc/actors/spacesuit_player.
I could not find where the NPC_SetLocomotionTargetSpeed is setup anywhere by default.  If it is at all.   
I'm going to take a wild guess that the NPC SetupLocomotionTargetSpeed goes on the characters as defined
the same archetype/npc/actors/ section for each character.   Neither of these settings exist on the characters stock.

If I want to setup the CHR_locomotion modifiers for the characters I'll need to manually define all the bindings and triggers.
I can use the spacesuit_player as a template for the others' settings.  Looks like it's triggered through finished spawning and 
finished respawning on each character as a function.   There's a problem though that Foster does not have this function 
called character as all the others do.  She's set up some other method.    I did however attempt to setup the other 
3 NPCs and there was no effect.   

I have no idea of the correct bindings and triggers to setup for the NPC SetupLocomotionTargetSpeed.  I tried duplcating 
the same bindings and triggers as the CHR_locomotionmodifier, but there was no effect.    

I think this is a blind alley.  I need to find something less involved.

Thank you for your help.  I'm just going to mess with some other way of doing this.

**Comment by ghramsey** [comment]
This whole saga gets weird now;  I figured out that the 2nd part of M9 does load after substantial delay.
Something close to a full minute for soem reason.

Anyway, I remain highly confused by this mission.  I reverted to a stock LV426_PT2 editing only the CHR Locomotion can_run modifier.
I saved. And loaded in to the mission.  It worked fine.  I exited the game and made on additional change.  I edited the delay 
for [F0-F9-AA-04] m13_pt2_scene_1 (CAGEAnimation) which I believe to be the final cut scene.  I changed it from 40 sec to 10.  
I saved and loaded up the game and into the mission. This small change breaks the characters' initial movement.  

I have to say this now W-T-F ?   Editing the delay back to 40 and saving does not undo the issue.  
So something is being saved and not un-done.   Not sure it's related, but saving the COMMANDS.PAK file
also saves a MODELS.MVR and REDS.BIN.  I understand the MVR is Mover but the latter file I've yet to 
encounter as such until now.  Is it perhaps related however remotely.
 If I load in the working file and save it w/o altering it, it no longer works when loaded in.   Maybe the MVR file?

**Comment by ghramsey** [comment]
With many thanks to JeffCat I have solved this problem with the frozen NPCs. 
Or rather Jeff made a custom GoTo movement binding to 'brute force' the NPCs' movement.   
While the source of the issue is unknown the movement problem is solved.

Thanks again JeffCat.

**Comment by ghramsey** [comment]
This is now fixed along with #417
